### PR TITLE
Add Cygwin GitHub Action.

### PR DESCRIPTION
**Pull request by hyoklee**
CMake test passes if `test/th5_system.c` has the following patch:

```
#ifdef __CYGWIN__
#define _GNU_SOURCE
#endif
```

Autotools test fails with the following error:

```
Testing MDC/FILE related API calls                                    *FAILED*
check_file_mdc_api_calls: failure_mssg = "Unexpected internal config #3.".
Testing MDC API smoke check                                            -SKIP-
     Long tests disabled.
Testing MDC/FILE related API input errors                             *FAILED*
check_file_mdc_api_errs: failure_mssg = "H5Fcreate() failed.
".
Testing MDC/FAPL related API calls for paged aggregation strategy     *FAILED*
check_fapl_mdc_api_calls: failure_mssg = "H5Fcreate() failed.
".
Testing MDC/FILE related API calls for paged aggregation strategy     *FAILED*
check_file_mdc_api_calls: failure_mssg = "H5Fcreate() failed.
".
Testing MDC API smoke check for paged aggregation strategy             -SKIP-
     Long tests disabled.
Testing MDC/FILE related API input errors for paged aggregation strategy*FAILED*
check_file_mdc_api_errs: failure_mssg = "H5Fcreate() failed.
```

### CODE REVIEW

**Comment by byrnHDF** [comment]
#define _GNU_SOURCE
That should be in ConfigueChecks.cmake and added to the global defines - not in code.

**Review by byrnHDF — COMMENTED** [review]
The yaml additions look good - Omly complaint is the name cwin-* maybe should be cygwin-*

**Comment by byrnHDF** [comment]
OH, yes tests may use different flags

File: configure.ac [hunk 1]
```diff
@@ -1519,6 +1519,10 @@ case "$host_cpu-$host_vendor-$host_os" in
     AC_DEFINE([HAVE_MINGW], [1], [Define if using MinGW])
     H5_CPPFLAGS="-D_GNU_SOURCE -D__USE_MINGW_ANSI_STDIO $H5_CPPFLAGS"
     ;;
+  *cygwin*)
+    AC_DEFINE([HAVE_CYGWIN], [1], [Define if using MinGW])
+    H5_CPPFLAGS="-D_GNU_SOURCE"
```
Comment (byrnHDF) [inline configure.ac]: Might need  $H5_CPPFLAGS so that it is appeneded (see the linux section above - line 1514)

**Review by byrnHDF — APPROVED** [review]

File: .github/workflows/cygwin-cmake.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+          -DCMAKE_BUILD_TYPE=${{ inputs.build_mode }} \
+          -D BUILD_SHARED_LIBS:BOOL=OFF \
+           ..
+        shell: C:\cygwin\bin\bash.exe -eo pipefail -o igncr '{0}'
+
+      - name: Build
+        run: |
+          export PATH=/usr/bin:$PATH
+          cd build
+          ctest -T Build
```
Comment (tbeu) [inline .github/workflows/cygwin-cmake.yml:42]:
```suggestion
          cmake --build build -- -j8
```

File: .github/workflows/cygwin-cmake.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        uses: cygwin/cygwin-install-action@master
+        with:
+          packages: cmake gcc-fortran make ninja perl zlib-devel
+
+      - name: Configure
+        run: |
+          export PATH=/usr/bin:$PATH
+          mkdir build
+          cd build
+          cmake -G Ninja \
```
Comment (tbeu) [inline .github/workflows/cygwin-cmake.yml:32]:
```suggestion
          cmake -S . -B build -G Ninja \
```

File: .github/workflows/cygwin-cmake.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+          export PATH=/usr/bin:$PATH
+          cd build
+          ctest -T Build
+        shell: C:\cygwin\bin\bash.exe -eo pipefail -o igncr '{0}'
+
+      - name: Test
+        run: |
+          export PATH=/usr/bin:$PATH
+          cd build
+          ctest -T Test
```
Comment (tbeu) [inline .github/workflows/cygwin-cmake.yml:49]:
```suggestion
          ctest --test-dir build -T Test
```

Comment (hyoklee) [reply .github/workflows/cygwin-cmake.yml:49]: @tbeu , thanks for taking the time to review my pull request! I also wanted to express my sincere gratitude for your work on maintaining the [matio](https://github.com/tbeu/matio) project. It's a valuable resource for the community. I'm happy to see that Cygwin is now a part of matio CI.

Comment (tbeu) [reply .github/workflows/cygwin-cmake.yml:49]: Sure, it's give and take. ❤️ 

(Cygwin via autotools was already tested in matio via Appveyor.)

**Comment by byrnHDF** [comment]
#4131 supersedes this PR - closing

**Pull request closed without merging**